toxcore/net_crypto.c

```diff
@@ -446,48 +525,206 @@ static int handle_cookie_response(const Logger *log, uint8_t *cookie, uint64_t *
+        NoiseBuffer noise_message;
+        //AKE NEW: ephemeral public key of the RESPONDER + encrypted payload + MAC of encrypted payload (with and without base nonce)
+//      uint8_t noise_message_buf[CRYPTO_PUBLIC_KEY_SIZE + CRYPTO_NONCE_SIZE + CRYPTO_SHA512_SIZE + COOKIE_LENGTH + CRYPTO_MAC_SIZE];
+        uint8_t noise_message_buf[CRYPTO_PUBLIC_KEY_SIZE + CRYPTO_SHA512_SIZE + COOKIE_LENGTH + CRYPTO_MAC_SIZE];
+        //AKE NEW: Application chosen payload of Noise message
+        NoiseBuffer noise_payload;
+        //AKE NEW: Payload contains NO base nonce, but sha512 hash of cookie (Initiator), and other cookie (Responder)
+//      uint8_t noise_payload_buf[CRYPTO_NONCE_SIZE + CRYPTO_SHA512_SIZE + COOKIE_LENGTH ];
+        uint8_t noise_payload_buf[CRYPTO_SHA512_SIZE + COOKIE_LENGTH ];
+        //AKE NEW: Add Nonce to payload => NOT necessary
+//      memcpy(noise_payload_buf, nonce, CRYPTO_NONCE_SIZE);
+        //AKE NEW: Add SHA512 of cookie to payload
+//      crypto_sha512(noise_payload_buf + CRYPTO_NONCE_SIZE, cookie, COOKIE_LENGTH);
+        crypto_sha512(noise_payload_buf, cookie, COOKIE_LENGTH);
+        uint8_t othercookie_plain[COOKIE_DATA_LENGTH ];
+        // AKE: The cookie contains the static and DHT pubkey of the peer! not from the sender!
+        memcpy(othercookie_plain, peer_real_pk, CRYPTO_PUBLIC_KEY_SIZE);
+        // AKE: The cookie contains the static and DHT pubkey of the peer! not from the sender!
+        memcpy(othercookie_plain + CRYPTO_PUBLIC_KEY_SIZE, peer_dht_pubkey, CRYPTO_PUBLIC_KEY_SIZE);
+
+        // AKE: This is to create the OTHER cookie from Peer A/initiator
+        // AKE: Also Peer B/responder sends a cookie again in the handshake packet
+        if (create_cookie(c->log, c->mono_time, noise_payload_buf + CRYPTO_SHA512_SIZE,
+                          othercookie_plain, c->secret_symmetric_key) != 0) {
+            return -1;
+        }
 
-    if (create_cookie(c->log, c->mono_time, plain + CRYPTO_NONCE_SIZE + CRYPTO_PUBLIC_KEY_SIZE + CRYPTO_SHA512_SIZE,
-                      cookie_plain, c->secret_symmetric_key) != 0) {
-        return -1;
-    }
+        //AKE NEW: Format the message payload into "payloadbuf".
+        noise_buffer_set_input(noise_payload, noise_payload_buf, sizeof(noise_payload_buf));
+        //AKE NEW: Points to the message buffer to be populated with handshake details and the message payload.
+        noise_buffer_set_output(noise_message, noise_message_buf, sizeof(noise_message_buf));
+        //AKE NEW: Computes Noise message B pattern "e, ee, se"
+        int err = noise_handshakestate_write_message(handshake, &noise_message, &noise_payload);
 
-    random_nonce(packet + 1 + COOKIE_LENGTH);
-    int len = encrypt_data(peer_real_pk, c->self_secret_key, packet + 1 + COOKIE_LENGTH, plain, sizeof(plain),
-                           packet + 1 + COOKIE_LENGTH + CRYPTO_NONCE_SIZE);
+        if (err != NOISE_ERROR_NONE) {
+            noise_perror("write handshake", err);
```

Comment:
This probably needs to become a LOGGER statement so we don't write to stderr from a library.